mantidproject/mantid
# Alter ISISReflectometryCalibration algorithm

**Description of work**

**Purpose of work**

A calibration step will be required when the new 2D detector starts being used on INTER in June. This calibration algorithm will also be useful for POLREF eventually.



**Summary of work**

I've re-written algorithm `ReflectometryISISCalibration` to implement the approach described in the referenced issue. Our Reflectometry scientists have a need to move individual detector pixels rather than moving the entire bank, as is sometimes done in other calibrations. They measure theta offsets for each pixel, which are then used to find the new detector pixel position. Existing algorithm `SpecularReflectionPositionCorrect` can calculate and apply the new positions, but internally uses method `MoveInstrumentComponent` to do this, which prevents moving individual pixels for rectangular (and other fixed) detectors, and affects INTER. After some discussion, we concluded that there wasn't a good science reason for this obstacle and so I have added a boolean parameter to `SpecularReflectionPositionCorrect` and `MoveInstrumentComponent` that allows the user to choose to force individual pixel moves for fixed detectors. The new parameter will default to preventing these moves to maintain backwards compatibility.

**Further detail of work**

The following tasks are still outstanding for this issue, but I would like to complete them a separate PR because of the need to get this merged in and tested by our users in time for the 6.7 release:

- Documentation updates - I need to add documentation for the `ReflectometryISISCalibration` algorithm. I will also review documentation for the `SpecularReflectionPositionCorrect` and `MoveInstrumentComponent` algorithms to see if anything needs updating after adding the new parameter. Finally, the docs for algorithm `ApplyCalibration` incorrectly state that it will not move rectangular detectors. This does not appear to be the case and so this should be removed.

- It would be good to add additional unit tests to cover more of the calibration file parsing logic. I would like to take a bit of time to see if there is a way of doing this that doesn't require lots of different test files to be provided.

- I need to add a unit test to cover the case where the algorithm is moving a detector pixel that has a negative theta value to start with.

- Issue #34906 still exists and would prevent POLREF using the algorithm via the GUI. Fixing this is not required for INTER, though, which is the current priority.

- If individual detector pixels are moved for a rectangular detector then Mantid currently prevents saving the calibrated workspace as a Nexus file. At the moment Reflectometry do not really use this functionality, but we will still need to find a solution to this eventually. This will be done under a separate issue - #35498.

**To test:**


1) Download this test INTER calibration file and remove the .txt extension (added just to be able to attach to GitHub): 
[INTER_calibration_full.dat.txt](https://github.com/mantidproject/mantid/files/11371051/INTER_calibration_full.dat.txt)
1) Open the ISIS Reflectometry GUI
1) In the runs table, enter 61135 under the Run column and 0.5 under Angle.
1) On the Experiment Settings tab, set the AnalysisMode to MultiDetectorAnalysis.
1) On the Instrument Settings tab, set the CalibrationFile path by browsing to the location of the downloaded calibration file.
1) Back on the Runs tab, click the Process button.
1) The reduction should complete successfully. If you right click on the TOF output workspace and choose Show Detectors you can view the theta values of each calibrated detector.
1) Use Load from the Workspaces pane in the main window to load the original INTER61135 data from the archive. Check the original theta values using Show Detector. Compare these to the theta values from the reduced, calibrated workspace. The theta values in the calibrated workspaces should be equal to original value + theta offset (taken from the test calibration file).
1) Check that changing the format of the test calibration file gives either an appropriate error message (i.e. providing invalid columns), or doesn't prevent the calibration from succeeding.

Refs #35352. 






---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
